jiant/proj/main/modeling/heads.py

```diff
@@ -1,29 +1,93 @@
+from __future__ import annotations
+
 import abc
 
 import torch
 import torch.nn as nn
 
 import transformers
+
 from jiant.ext.allennlp import SelfAttentiveSpanExtractor
+from jiant.shared.model_resolution import ModelArchitectures
+from jiant.tasks.core import TaskTypes
+from typing import Callable
+from typing import List
 
 """
 In HuggingFace/others, these heads differ slightly across different encoder models.
 We're going to abstract away from that and just choose one implementation.
 """
 
 
+class JiantHeadFactory:
+    """This factory is used to create task-specific heads for the supported Transformer encoders.
+
+    Attributes:
+        registry (dict): Dynamic registry mapping task types to task heads
+    """
+
+    registry = {}
+
+    @classmethod
+    def register(cls, task_type_list: List[TaskTypes]) -> Callable:
+        """Register each TaskType in task_type_list as a key mapping to a BaseHead task head
+
+        Args:
+            task_type_list (List[TaskType]): List of TaskTypes that are associated to a
+                                             BaseHead task head
+
+        Returns:
+            Callable: inner_wrapper() wrapping task head constructor or task head factory
+        """
+
+        def inner_wrapper(wrapped_class: BaseHead) -> Callable:
+            """Summary
+
+            Args:
+                wrapped_class (BaseHead): Task head class
+
+            Returns:
+                Callable: Task head constructor or factory
+            """
+            for task_type in task_type_list:
+                assert task_type not in cls.registry
+                cls.registry[task_type] = wrapped_class
+            return wrapped_class
+
+        return inner_wrapper
+
+    def __call__(self, task, **kwargs) -> BaseHead:
+        """Summary
+
+        Args:
+            task (Task): A task head will be created based on the task type
+            **kwargs: Arguments required for task head initialization
+
+        Returns:
+            BaseHead: Initialized task head
+        """
+        head_class = self.registry[task.TASK_TYPE]
+        head = head_class(task, **kwargs)
+        return head
+
+
 class BaseHead(nn.Module, metaclass=abc.ABCMeta):
-    pass
+    """Absract class for task heads"""
+
+    @abc.abstractmethod
+    def __init__(self):
+        super().__init__()
 
 
+@JiantHeadFactory.register([TaskTypes.CLASSIFICATION])
 class ClassificationHead(BaseHead):
-    def __init__(self, hidden_size, hidden_dropout_prob, num_labels):
+    def __init__(self, task, hidden_size, hidden_dropout_prob, **kwargs):
```

Comment:
Could you remind me why the task implementations require `**kwargs`?

Reply:
This is a result of using the `JiantHeadFactory` in model_setup.py for MLM heads. The following snippet summarizes how it is used.

```python
def create_taskmodel(task, encoder, **taskmodel_kwargs) -> Taskmodel:
    """Creates, initializes and returns the task model for a given task type and encoder.

    Args:
        task (Task): Task object associated with the taskmodel being created.
        encoder (JiantTransformersModel): Transformer w/o heads
            (embedding layer + self-attention layer).
        **taskmodel_kwargs: Additional args for taskmodel setup

    Raises:
        KeyError if task does not have valid TASK_TYPE.

    Returns:
        Taskmodel

    """
    head_kwargs = {}
    head_kwargs["hidden_size"] = encoder.get_hidden_size()
    head_kwargs["hidden_dropout_prob"] = encoder.get_hidden_dropout_prob()
    head_kwargs["vocab_size"] = encoder.config.vocab_size
    head_kwargs["model_arch"] = ModelArchitectures(encoder.config.model_type)

    if hasattr(encoder, "hidden_act"):
        head_kwargs["hidden_act"] = encoder.config.hidden_act
    if hasattr(encoder, "layer_norm_eps"):
        head_kwargs["layer_norm_eps"] = encoder.config.layer_norm_eps

    head = JiantHeadFactory()(task, **head_kwargs)

    taskmodel = JiantTaskModelFactory()(task, encoder, head, **taskmodel_kwargs)
    return taskmodel
```

The kwargs are used is the MLM heads such as the following: 

```python
@JiantMLMHeadFactory.register([ModelArchitectures.ROBERTA, ModelArchitectures.XLM_ROBERTA])
class RobertaMLMHead(BaseMLMHead):
    """From RobertaLMHead"""

    def __init__(self, hidden_size, vocab_size, layer_norm_eps=1e-12):
        super().__init__()
        self.dense = nn.Linear(hidden_size, hidden_size)
        self.layer_norm = transformers.models.bert.modeling_bert.BertLayerNorm(
            hidden_size, eps=layer_norm_eps
        )

        self.decoder = nn.Linear(hidden_size, vocab_size, bias=False)
        self.bias = nn.Parameter(torch.zeros(vocab_size), requires_grad=True)

        # Need a link between the two variables so that the bias is correctly resized with
        # `resize_token_embeddings`
        self.decoder.bias = self.bias
```

Reply:
I believe this was previously passed around as an explicit dictionary rather than a **kwargs.

That said, I'm happy to tweak this after merging.